Add priority to SHFlow, replace _flowPool with flat_set

sinkingsugar opened the pull request:
<!--
Before submitting your PR, please review the following checklist:

- [ ] **DO NOT** submit a PR without having discussed the change first in a related issue. If none exist, create one first.
- [ ] **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] **DO** keep pull requests small so they can be easily reviewed.
- [ ] **DO** make sure all unit tests pass.
- [ ] **DO** make sure not to introduce any new compiler warnings.
- [ ] **AVOID** breaking the continuous integration build.
- [ ] **AVOID** making significant changes to the overall architecture.
-->

sinkingsugar committed:
Add priority to SHFlow, replace _flowPool with flat_set

sinkingsugar committed:
Prevent duplicate wire scheduling in SHMesh flow setup

Check for duplicate wire scheduling in `_flowPool.insert`. If insertion fails due to an existing entry, log an error and throw an `SHException` with a descriptive message, ensuring wires are only scheduled once.

sinkingsugar committed:
Fix wireCleanedUp

sinkingsugar committed:
Add priority support for SHWire in scheduler and tests

Introduce `setWirePriority` function and `priority` field to `SHWire` for controlling scheduling priority in the runtime. Update flow creation to pass priority and clamp the value within integer bounds. Modify `network.shs` tests to assign priorities, ensuring priority-based execution consistency across test cases.

sinkingsugar committed:
Refactor network test shards by removing priority tags

This commit removes all `Priority` tags from various network test shards, including `network-test-server`, `client-init`, and `network-test-client` configurations. Adjustments ensure cleaner scheduling without explicit priority settings, enhancing maintainability and alignment with current shard practices.

sinkingsugar committed:
Refactor network-ws.shs to modularize main test

Refactor network-ws.shs to use a new @wire and @schedule approach for running tests with Network.WS client and server. Replace Network.Client and Network.Server with WebSocket-based versions in main-test wire.

sinkingsugar committed:
Make sure server starts first in network tests

sinkingsugar committed:
Refactor `SHFlow` initialization, enhance cleanup logging

Refactor `SHFlow` structure initialization by adding a constructor with `priority`, `wire`, and `paused` arguments, and implement a less-than operator to handle priority-based sorting in `_flowPool`. Remove redundant `SHFlowLess` struct, streamline flow pool removal with improved iterator handling, and add trace logging for wire cleanup in `SHMesh`. Enhance serialization by adding `priority` to wire read/write operations and fix formatting in audio and genetic modules.

sinkingsugar committed:
Fix mesh terminate

sinkingsugar committed:
Proper order for resumer restoring

sinkingsugar committed:
Refactor `SHFlow` to encapsulate paused and wire state, streamline flow swapping in mesh, and ensure resumption logic consistency

This commit introduces improvements to `SHFlow` by encapsulating the paused state and wire pointer, allowing for getter and setter methods (`isPaused`, `setPaused`, `getWire`). Flow swapping logic is centralized in a new `swapFlows` function within `SHMesh`, ensuring consistent handling of flow priority and proper resumption behavior in cases where a resumer is set. This update also adjusts `SuspendWire` and `ResumeWire` to use the encapsulated paused state, and refactors logic in `wireCleanedUp` to manage flow removal accurately. Finally, `SwitchTo` updates `context->flow` using the mesh swap for better flow state management across ticks and wire transitions.

sinkingsugar committed:
Refactor SHWire event handling and cleanup process

Refactor SHWire's event handling by centralizing dispatcher triggering within each wire instance instead of relying on the mesh. Add scoped connections for OnStart and OnCleanup events, enhancing safety and automatic disconnection upon destruction. Replace `scheduled` checks with direct removal from `_flowPool` during cleanup. Simplify `tick` function flow by updating iterator handling, removing unused flow state assertions, and eliminating redundant `wireCleanedUp` function from SHMesh. Refactor `Spawn` and `WhenDone` to handle cleanup connections using `scoped_connection` for safer lifecycle management.

sinkingsugar committed:
Refactor: simplify cleanup connections in Spawn and WhenDone

Remove unnecessary `_connectedMesh` member and related checks in Spawn and WhenDone. Establish scoped connections for each `ManyWire` instance to handle cleanup events, reducing redundancy and ensuring that each wire has its own scoped connection.

sinkingsugar committed:
progress on removing shflow

sinkingsugar committed:
Nearly there, simplified a lot

sinkingsugar committed:
Enhance wire handling with child wire delegation in SwitchTo

Refactor `SHWire` to handle nested wire delegation with `childWire`, allowing `SwitchTo` operations to seamlessly transition control between wires. Add `tickingWire()` function to retrieve the active wire context, updating calls to `tick()` accordingly. Ensure proper child wire cleanup by resetting `childWire` and `resumer` pointers post-switch. Add validation and scheduling for non-started wires and enhance error logging for wire failures. Adjust logging and unscheduling logic to reflect expected `SwitchTo` behavior.

sinkingsugar committed:
Final touches with SwitchTo and Step

guusw commented on shards/tests/network-ws.shs:
```diff
@@ -1,12 +1,17 @@
-
 @mesh(root)
 
+@wire(main-test {
```
Why can't this run on the script root anymore, WireRunner doing something?

sinkingsugar replied:
should be ok actually, will try revert

sinkingsugar committed:
Improve wire management and ID regeneration logic

- Update `SHWire` to use `uniqueId` for consistent ordering in `SHMesh`
- Add `regenerateId` to ensure wires are added at the end of the mesh list
- Modify `tickingWire` function to traverse nested `childWire` correctly
- Add assertions in `WireDoppelgangerPool` for wire integrity checks
- Include `allCleared` function in `SHMesh` to verify all containers are empty
- Log wire creation and ensure ID regeneration

sinkingsugar committed:
Revert network test

sinkingsugar committed:
Remove SHFlow struct from runtime and update SwitchTo
